### DISCUSSION TITLE
MediatR abtraction

### DISCUSSION THREAD

**Discussion by mahmoudshalsh** [LuckyPennySoftware/MediatR#725] [Q&A]
@jbogard your advice, please
I need to depend on my custom interface with dotnet 6
example:

```
using Microsoft.AspNetCore.Mvc;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

builder.Services.AddSingleton<ServiceFactory>(serviceProvider => (type) => serviceProvider.GetRequiredService(type));
builder.Services.AddScoped<IMediator, Mediator>();
builder.Services.AddScoped<IQueryHandler<Query, Result>, Handler>();

var app = builder.Build();

app.UseHttpsRedirection();
app.UseSwagger();
app.UseSwaggerUI();

app.MapGet("/", ([FromServices] IMediator mediator, [FromQuery] string name) => mediator.Send(new Query(name)));

app.Run();

public record Query(string Name) : IQuery<Result>;

public record Result(string Message);

public class Handler : IQueryHandler<Query, Result>
{
    public Task<Result> Handle(Query query, CancellationToken cancellationToken)
        => Task.FromResult(new Result($"Hello {query.Name}"));
}

public interface IQuery<out TResponse> { }

public interface IQueryHandler<in TQuery, TResult> where TQuery : IQuery<TResult>
{
    Task<TResult> Handle(TQuery query, CancellationToken cancellationToken);
}

```

**Comment by jbogard** [comment]
I don't understand, what is the question here? You don't want to use the built-in interface with MediatR? Why?

**Reply by mahmoudshalsh** [reply]
@jbogard 
We write all requests, notifications and handers in Application Service layer , we need to abstract this layer from any technologies dependencies like MediatR NuGet package.
So we need to tell mediator class how to resolve our custom handler

**Reply by jbogard** [reply]
Why not just use the Contracts assembly? It has only the interfaces.

**Reply by jbogard** [reply]
Ah, it only has the request/response/notification contract interfaces. Not the handlers.

If you're making this architectural choice to "not have dependencies", then you'll need to copy the code from inside MediatR to your application. Or just reference MediatR directly, as the line of what is a "dependency" or not is completely arbitrary. Everything you're writing takes a dependency on SOMETHING in order to run.
